topology/probes/libvirt/libvirt.go

```diff
@@ -214,85 +215,119 @@ func (probe *Probe) deleteDomain(d *libvirtgo.Domain) {
+	probe.cidLifecycle, err = probe.conn.DomainEventLifecycleRegister(nil, probe.onLibvirtDomainEvent)
 	if err != nil {
 		logging.GetLogger().Errorf(
 			"Could not register the lifecycle event handler %s", err)
+		return false
 	}
-	callbackDeviceAdded := func(
-		c *libvirtgo.Connect, d *libvirtgo.Domain,
-		event *libvirtgo.DomainEventDeviceAdded,
-	) {
-		domainNode := probe.getDomain(d)
-		interfaces := probe.getDomainInterfaces(d, domainNode, event.DevAlias)
-		probe.registerInterfaces(interfaces) // 0 or 1 device changed.
-	}
-	probe.cidDevAdded, err = conn.DomainEventDeviceAddedRegister(nil, callbackDeviceAdded)
+	probe.cidDevAdded, err = probe.conn.DomainEventDeviceAddedRegister(nil, probe.onLibvirtDeviceEvent)
 	if err != nil {
 		logging.GetLogger().Errorf(
 			"Could not register the device added event handler %s", err)
+		return false
 	}
 	domains, err := probe.conn.ListAllDomains(0)
 	if err != nil {
 		logging.GetLogger().Error(err)
-		return
+		return false
 	}
 	for _, domain := range domains {
 		domainNode := probe.createOrUpdateDomain(&domain)
 		interfaces := probe.getDomainInterfaces(&domain, domainNode, "")
 		probe.registerInterfaces(interfaces)
 	}
+
+	return true
+}
+
+func (probe *Probe) onLibvirtDomainEvent(c *libvirtgo.Connect, d *libvirtgo.Domain, event *libvirtgo.DomainEventLifecycle) {
+	switch event.Event {
+	case libvirtgo.DOMAIN_EVENT_UNDEFINED:
+		probe.deleteDomain(d)
+	case libvirtgo.DOMAIN_EVENT_STARTED:
+		domainNode := probe.createOrUpdateDomain(d)
+		interfaces := probe.getDomainInterfaces(d, domainNode, "")
+		probe.registerInterfaces(interfaces)
+	case libvirtgo.DOMAIN_EVENT_DEFINED, libvirtgo.DOMAIN_EVENT_SUSPENDED,
+		libvirtgo.DOMAIN_EVENT_RESUMED, libvirtgo.DOMAIN_EVENT_STOPPED,
+		libvirtgo.DOMAIN_EVENT_SHUTDOWN, libvirtgo.DOMAIN_EVENT_PMSUSPENDED,
+		libvirtgo.DOMAIN_EVENT_CRASHED:
+		probe.createOrUpdateDomain(d)
+	}
+}
+
+func (probe *Probe) onLibvirtDeviceEvent(
+	c *libvirtgo.Connect, d *libvirtgo.Domain,
+	event *libvirtgo.DomainEventDeviceAdded,
+) {
+	domainNode := probe.getDomain(d)
+	interfaces := probe.getDomainInterfaces(d, domainNode, event.DevAlias)
+	probe.registerInterfaces(interfaces) // 0 or 1 device changed.
+}
+
+func (probe *Probe) periodicChecker(ctx context.Context) {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			{
+				ticker.Stop()
+				if probe.cidLifecycle != -1 {
+					probe.conn.DomainEventDeregister(probe.cidLifecycle)
+					probe.conn.DomainEventDeregister(probe.cidDevAdded)
+				}
+				probe.conn.Close()
+				return
+			}
+		case <-ticker.C:
+			{
+				var err error
+				var ok bool
+				if probe.conn == nil {
+					ok = probe.tryToConnectToLibvirt()
+					if !ok {
+						continue
+					}
+				}
+				if ok, err = probe.conn.IsAlive(); err != nil || !ok {
+					ok = probe.tryToConnectToLibvirt()
+					if !ok {
+						continue
+					}
+				}
+				if err = libvirtgo.EventRunDefaultImpl(); err != nil {
```

Comment:
This is going to be called only every 5 sec ? it should be called in a loop with no delay  no ?

Reply:
fixed.